### DISCUSSION TITLE
AssertionError: cycle-free graph violated: partition depends on itself

### DISCUSSION THREAD

**Discussion by danielflood** [Xilinx/finn#495] [Q&A] [answered]
Hey, 
I'm trying to use the build dataflow tool on my own custom brevitas model by following the steps in the cybersecurity notebook.
```
  cfg_estimates = build.DataflowBuildConfig(
      output_dir          = estimates_output_dir,
      mvau_wwidth_max     = 80,
      target_fps          = 1000000,
      synth_clk_period_ns = 10.0,
      fpga_part           = "xc7z020clg400-1",
      steps               = build_cfg.estimate_only_dataflow_steps,
      generate_outputs=[
          build_cfg.DataflowOutputType.ESTIMATE_REPORTS,
      ]
  )
```
However, when I run``` build.build_dataflow_cfg(model_file, cfg_estimates)```, it gets stuck on the to_hls step with the following error:
**AssertionError: cycle-free graph violated: partition depends on itself**

I was able to find the node that is causing this error which is the StreamingFCLayer_Batch node which has an input of shape of 1x98x1x500 and an output of shape 1x98x1x500 as can be seen in the graph below. This graph is the intermediate output from to_hls step when building the dataflow architecture.

Can anyone see the cause of this issue from the graph or able offer any advice on how to debug this issue?

Thanks,
Dan

![step_convert_to_hls onnx(1)](https://user-images.githubusercontent.com/62947583/149543420-b6a3f8ad-87ab-4c26-8716-b6e3ad9ba7f9.png)

**Comment by nhma20** [comment, marked as answer]
I believe this can happen if the feature dimension going into a maxpool layer (size=2, stride=2) is not divisible by 2:
Input image size 32x32:   32 (padded conv2d) 32 (maxpool) 16 (padded conv2d) 16 (maxpool) 8 ... (Good)
Input image size 30x30:   30 (padded conv2d) 30 (maxpool) 15 (padded conv2d) 15 (maxpool) 7? (Bad?)

**Reply by danielflood** [reply]
Hey sorry only seeing this now.
Yep! That was exactly the problem. Changing the size of my max pool resolved it.
Thanks for your help!

**Comment by Govarshini99** [comment]
I have a similar issue with my model but i have no maxpool layers
![fish_mobilenet_v2_tf1 onnx](https://github.com/user-attachments/assets/5674b78a-a1fd-4c3f-b84b-4f6024bc0128)

this is my model

**Reply by lucas-coutinho** [reply]
I think your problem is due to the Im2Col layers that were not converted to hardware properly. And probably is due to the Mul layers that cast the input to Im2Col to float, which skips the step when converting to hw (Im2Col are converted only if the input datatype is int).
